pkg/fqdn/service/service.go

```diff
@@ -0,0 +1,202 @@
+	ctx := context.Background()
+
+	s := &FQDNDataServer{
+		endpointManager:     endpointManager,
+		updateOnDNSMsg:      updateOnDNSMsg,
+		ctx:                 ctx,
+		currentIdentityToIp: make(map[identity.NumericIdentity][]net.IP),
+		log:                 logger.With(logfields.LogSubsys, "fqdn/server"),
+	}
+
+	go func() {
+		<-s.ctx.Done()
+		s.log.Info("FQDN service context done, cleaning up resources")
+		if s.ctx.Err() != nil {
+			s.log.Error("FQDN service context error: ", logfields.Error, s.ctx.Err())
+		}
+		s.closeServerMu.Lock()
+		closeFunc := s.closeServer
+		s.closeServerMu.Unlock()
+		if closeFunc != nil {
+			closeFunc()
+		}
+	}()
+
+	return s
+}
+
+// OnIPIdentityCacheChange is a method to receive the IP identity cache change events
+func (s *FQDNDataServer) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr types.PrefixCluster, oldHostIP, newHostIP net.IP, oldID *ipcache.Identity, newID ipcache.Identity, encryptKey uint8, k8sMeta *ipcache.K8sMetadata, endpointFlags uint8) {
+	s.identityToIpMutex.Lock()
+	defer s.identityToIpMutex.Unlock()
+
+	if cidr.ClusterID() == 0 {
+
+		ipNet := cidr.AsIPNet()
+		ones, _ := ipNet.Mask.Size()
+
+		if (ipNet.IP.To4() != nil && ones != 32) || (ipNet.IP.To4() == nil && ones != 128) {
+			s.log.Info("CIDR mask not supported", logfields.CIDR, cidr.String())
+		} else {
+			switch modType {
+			case ipcache.Upsert:
+				if oldID != nil {
+					// Remove from the old identity
+					if ips, ok := s.currentIdentityToIp[oldID.ID]; ok {
+						for i, ip := range ips {
+							if ip.Equal(ipNet.IP) {
+								s.currentIdentityToIp[oldID.ID] = append(ips[:i], ips[i+1:]...)
+								break
+							}
+						}
+					}
+				}
+				s.currentIdentityToIp[newID.ID] = append(s.currentIdentityToIp[newID.ID], ipNet.IP)
+
+			case ipcache.Delete:
+				if oldID != nil {
+					if ips, ok := s.currentIdentityToIp[oldID.ID]; ok {
+						for i, ip := range ips {
+							if ip.Equal(ipNet.IP) {
+								s.currentIdentityToIp[oldID.ID] = append(ips[:i], ips[i+1:]...)
+								break
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+}
+
+// UpdatePolicyRulesLocked updates the current state of the DNS rules with the given policies and sends the current state of the DNS rules to the client
+// This method is called:
+// 1. when the DNS rules are updated during the endpoint regeneration, we store the state of the DNS rules with flag rulesUpdate as true
+// 2. when the client subscribes to DNS policies, we send the current state of the DNS rules to the client(flag rulesUpdate as false)
+// 3. when the IP identity cache changes, we update the current state of the identity to IP mapping and send the current state of the DNS rules to
+// the client(flag rulesUpdate as false)
+func (s *FQDNDataServer) UpdatePolicyRulesLocked(policies map[identity.NumericIdentity]policy.SelectorPolicy, rulesUpdate bool) error {
```

Comment:
Cilium convention is that any method with the suffix `Locked` indicates that it must be called with a particular lock held. Could you document which lock must be held here? It is not clear to me from context.

Reply:
Added the comment for proxy lock to be taken while we call this.